haystack/preview/components/builders/dynamic_prompt_builder.py

```diff
@@ -0,0 +1,334 @@
+        def run(self, doc_input: str):
+            return {"documents": [Document(content=doc_input)]}
+
+
+
+    pipe = Pipeline()
+    pipe.add_component("doc_producer", DocumentProducer())
+    pipe.add_component("prompt_builder", prompt_builder)
+    pipe.add_component("llm", llm)
+
+    # note here how prompt_builder.documents is received from doc_producer.documents
+    pipe.connect("doc_producer.documents", "prompt_builder.documents")
+    pipe.connect("prompt_builder.prompt", "llm.messages")
+
+    messages = [ChatMessage.from_system("Be helpful assistant, but brief!"),
+                ChatMessage.from_user("Here is the document: {{documents[0].content}} Now, answer the
+                following: {{query}}")]
+
+
+    pipe.run(data={"doc_producer": {"doc_input": "Hello world, I'm Haystack!"},
+                   "prompt_builder": {"prompt_source": messages,
+                                      "template_variables":{"query": "who's making a greeting?"}}})
+
+    >> {'llm': {'replies': [ChatMessage(content='Haystack', role=<ChatRole.ASSISTANT: 'assistant'>, name=None,
+    >> metadata={'model': 'gpt-3.5-turbo-0613', 'index': 0, 'finish_reason': 'stop', 'usage':
+    >> {'prompt_tokens': 51, 'completion_tokens': 2, 'total_tokens': 53}})]}}
+    ```
+
+    Similarly to chat prompt generation, you can use DynamicPromptBuilder to generate non-chat-based prompts.
+    The following example demonstrates how to use DynamicPromptBuilder to generate a non-chat prompt:
+
+    ```python
+    prompt_builder = DynamicPromptBuilder(expected_runtime_variables=["documents"], chat_mode=False)
+    llm = GPTGenerator(api_key="<your-api-key>", model_name="gpt-3.5-turbo")
+
+
+    @component
+    class DocumentProducer:
+
+      @component.output_types(documents=List[Document])
+      def run(self, doc_input: str):
+        return {"documents": [Document(content=doc_input)]}
+
+
+    pipe = Pipeline()
+    pipe.add_component("doc_producer", DocumentProducer())
+    pipe.add_component("prompt_builder", prompt_builder)
+    pipe.add_component("llm", llm)
+    pipe.connect("doc_producer.documents", "prompt_builder.documents")
+    pipe.connect("prompt_builder.prompt", "llm.prompt")
+
+    template = "Here is the document: {{documents[0].content}} \n Answer: {{query}}"
+    pipe.run(data={"doc_producer": {"doc_input": "Hello world, I live in Berlin"},
+               "prompt_builder": {"prompt_source": template,
+                                  "template_variables":{"query": "Where does the speaker live?"}}})
+
+    >> {'llm': {'replies': ['The speaker lives in Berlin.'],
+    >> 'metadata': [{'model': 'gpt-3.5-turbo-0613',
+    >> 'index': 0,
+    >> 'finish_reason': 'stop',
+    >> 'usage': {'prompt_tokens': 28,
+    >> 'completion_tokens': 6,
+    >> 'total_tokens': 34}}]}}
+
+    """
+
+    def __init__(self, expected_runtime_variables: Optional[List[str]] = None, chat_mode: Optional[bool] = True):
```

Comment:
I would change `expected_runtime_variables` to just `runtime_variables`. It seems a bit redundant to prefix it with `expected_`, what do you think?

Reply:
I'm ok with both, IIRC correctly @masci and I talked about `expected_variables` or `expected_runtime_variables` I didn't write it down tbh.